packages/vitest/src/node/reporters/junit.ts

```diff
@@ -0,0 +1,198 @@
+
+      // reached at test file, skip the follow stack
+      if (frame.file in this.ctx.state.filesMap)
+        break
+    }
+  }
+
+  async logLogs(task: Task, type: 'err' | 'out'): Promise<void> {
+    if (task.logs == null || task.logs.length === 0)
+      return
+
+    const logType = type === 'err' ? 'stderr' : 'stdout'
+    const logs = task.logs.filter(log => log.type === logType)
+
+    if (logs.length > 0) {
+      await this.logger.log(`<system-${type}>`)
+      for (const log of logs)
+        await this.baseLog(escapeXML(log.content))
+      await this.logger.log(`</system-${type}>`)
+    }
+  }
+
+  async writeTasks(tasks: Task[], filename: string): Promise<void> {
+    for (const task of tasks) {
+      await this.logger.log(openTag('testcase', {
+        classname: filename,
+        name: task.name,
+        time: getDuration(task),
+      }))
+
+      this.logger.indent()
+
+      await this.logLogs(task, 'out')
+      await this.logLogs(task, 'err')
+
+      if (task.mode === 'skip' || task.mode === 'todo')
+        this.logger.log('<skipped/>')
+
+      if (task.result?.state === 'fail') {
+        const error = task.result.error
+
+        await this.logger.log(openTag('failure', {
+          message: error?.message,
+          type: error?.name ?? error?.nameStr,
+        }))
+
+        if (error)
+          await this.logError(error)
+
+        await this.logger.log('</failure>')
+      }
+
+      this.logger.unindent()
+
+      await this.logger.log('</testcase>')
+    }
+  }
+
+  async onFinished(files = this.ctx.state.getFiles()) {
+    await this.logger.log('<?xml version="1.0" encoding="UTF-8" ?>')
+
+    const transformed = files
+      .map((file) => {
+        const tasks = file.tasks.flatMap(task => flattenTasks(task))
+
+        const stats = tasks.reduce((stats, task) => {
+          return {
+            passed: stats.passed + Number(task.result?.state === 'pass'),
+            failures: stats.failures + Number(task.result?.state === 'fail'),
+            skipped: stats.skipped + Number(task.mode === 'skip' || task.mode === 'todo'),
+          }
+        },
+        {
+          passed: 0,
+          failures: 0,
+          skipped: 0,
+        })
+
+        return {
+          ...file,
+          stats,
+        }
+      })
+
+    await this.logger.log('<testsuites>')
+    this.logger.indent()
+
+    for (const file of transformed) {
+      await this.logger.log(openTag('testsuite', {
+        name: file.name,
+        timestamp: (new Date()).toISOString(),
+        hostname: hostname(),
+        tests: file.tasks.length,
+        failures: file.stats.failures,
+        errors: 0, // An errored test is one that had an unanticipated problem. We cannot detect those.
+        skipped: file.stats.skipped,
+        time: getDuration(file),
+      }))
+
+      this.logger.indent()
+      await this.writeTasks(file.tasks, file.name)
+      this.logger.unindent()
+
+      await this.logger.log('</testsuite>')
```

Comment:
Would it make sense to have a `logElement` function that will abstract the indent, unindent and closing of tags to simplify this reporter?
```ts
  logElement: async function(name: string, attrs: Record<string, any>, children: () => void) {
name: string, attrs: Record<string, any>) {
    await this.logger.log(openTag(name, attrs))
    this.logger.indent()
    await children.call(this)
    this.logger.unindent()

    await this.logger.log(`</${name}>`)
  }
```
So you could call `logElement` instead here (and in other places where you have `openTag`, indent, etc
```ts
      await this.logElement('testsuite', {
        name: file.name,
        timestamp: (new Date()).toISOString(),
        hostname: hostname(),
        tests: file.tasks.length,
        failures: file.stats.failures,
        errors: 0, // An errored test is one that had an unanticipated problem. We cannot detect those.
        skipped: file.stats.skipped,
        time: getDuration(file),
      }, () => {
        await this.writeTasks(file.tasks, file.name)
      })
```

Reply:
Makes sense. Will implement this refactoring tomorrow.